ocaml/xapi/xapi_observer.ml

```diff
@@ -252,92 +368,79 @@ let unregister ~__context ~self ~host:_ =
 
-let set_max_spans ~__context spans =
-  List.iter
-    (fun c ->
-      let module Forwarder = (val get_forwarder c : ObserverInterface) in
-      Forwarder.set_max_spans ~__context ~spans
-    )
-    supported_components
+let set_max_spans ~__context spans component =
+  let module Forwarder = (val get_forwarder component : ObserverInterface) in
+  Forwarder.set_max_spans ~__context ~spans
 
-let set_max_traces ~__context traces =
-  List.iter
-    (fun c ->
-      let module Forwarder = (val get_forwarder c : ObserverInterface) in
-      Forwarder.set_max_traces ~__context ~traces
-    )
-    supported_components
+let set_max_traces ~__context traces component =
+  let module Forwarder = (val get_forwarder component : ObserverInterface) in
+  Forwarder.set_max_traces ~__context ~traces
 
-let set_max_file_size ~__context file_size =
-  List.iter
-    (fun c ->
-      let module Forwarder = (val get_forwarder c : ObserverInterface) in
-      Forwarder.set_max_file_size ~__context ~file_size
-    )
-    supported_components
+let set_max_file_size ~__context file_size component =
+  let module Forwarder = (val get_forwarder component : ObserverInterface) in
+  Forwarder.set_max_file_size ~__context ~file_size
 
-let set_host_id ~__context host_id =
-  List.iter
-    (fun c ->
-      let module Forwarder = (val get_forwarder c : ObserverInterface) in
-      Forwarder.set_host_id ~__context ~host_id
-    )
-    supported_components
+let set_host_id ~__context host_id component =
+  let module Forwarder = (val get_forwarder component : ObserverInterface) in
+  Forwarder.set_host_id ~__context ~host_id
 
-let set_compress_tracing_files ~__context enabled =
-  List.iter
-    (fun c ->
-      let module Forwarder = (val get_forwarder c : ObserverInterface) in
-      Forwarder.set_compress_tracing_files ~__context ~enabled
-    )
-    supported_components
+let set_compress_tracing_files ~__context enabled component =
+  let module Forwarder = (val get_forwarder component : ObserverInterface) in
+  Forwarder.set_compress_tracing_files ~__context ~enabled
 
-let init ~__context =
-  List.iter
-    (fun c ->
-      let module Forwarder = (val get_forwarder c : ObserverInterface) in
-      Forwarder.init ~__context
-    )
-    supported_components
+let init ~__context component =
+  let module Forwarder = (val get_forwarder component : ObserverInterface) in
+  Forwarder.init ~__context
 
-let load ~__context =
+let load ~__context component =
   let all = Db.Observer.get_all ~__context in
   List.iter
     (fun self ->
       let host = Helpers.get_localhost ~__context in
       let hosts = Db.Observer.get_hosts ~__context ~self in
       if hosts = [] || List.mem host hosts then
-        register ~__context ~self ~host
+        register_component ~__context ~self ~host ~component
     )
     all
 
+let initialise_component ~__context =
+  let comp_left = ref (List.map (fun c -> (c, 1)) supported_components) in
```

Comment:
The use of `comp_left` is a bit cryptic... It is essentially doing an intersection between the set of "supported components" and "observed components"? If so, then I think it'd be clearer to literally do that set operation in `initialise` below.

Reply:
The `Db.Observer.get_all` function returns a list of observers, each of which might has its own associated component. These components might overlap, e.g. [xapi], [xapi, xenopsd]. In this case, we want to make sure we call the initialisation functions liek `set_trace_log_dir` once for xapi and xenopsd.

Reply:
However, it is not the case for the `load` function which needs to be called once for *every* component. E.g. with the example above we need to call `load` for `xapi` twice, and once for xenopsd.

Reply:
Although I don't think it is crucially important here as most of the functions like `set_trace_log_dir` is (sort of) idempotent. Although the init function is not, because it will create an exporting thread each time it is called, so we need to be careful with this.